packages/material-ui/src/ButtonBase/focusVisible.js

```diff
@@ -1,75 +1,111 @@
+    return true;
+  }
+
+  if (tagName === 'TEXTAREA' && !node.readOnly) {
+    return true;
   }
-  return activeElement;
+
+  if (node.isContentEditable) {
+    return true;
+  }
+
+  return false;
 }
 
-export function detectFocusVisible(instance, element, callback, attempt = 1) {
-  warning(instance.focusVisibleCheckTime, 'Material-UI: missing instance.focusVisibleCheckTime.');
-  warning(
-    instance.focusVisibleMaxCheckTimes,
-    'Material-UI: missing instance.focusVisibleMaxCheckTimes.',
-  );
+function handleKeyDown() {
+  hadKeyboardEvent = true;
+}
 
-  instance.focusVisibleTimeout = setTimeout(() => {
-    const doc = ownerDocument(element);
-    const activeElement = findActiveElement(doc);
+/**
+ * If at any point a user clicks with a pointing device, ensure that we change
+ * the modality away from keyboard.
+ * This avoids the situation where a user presses a key on an already focused
+ * element, and then clicks on a different element, focusing it with a
+ * pointing device, while we still think we're in keyboard modality.
+ * @param {Event} e
+ */
+function handlePointerDown() {
+  hadKeyboardEvent = false;
+}
 
-    if (
-      internal.focusKeyPressed &&
-      (activeElement === element || element.contains(activeElement))
-    ) {
-      callback();
-    } else if (attempt < instance.focusVisibleMaxCheckTimes) {
-      detectFocusVisible(instance, element, callback, attempt + 1);
+function handleVisibilityChange() {
+  if (document.visibilityState === 'hidden') {
+    // If the tab becomes active again, the browser will handle calling focus
+    // on the element (Safari actually calls it twice).
+    // If this tab change caused a blur on an element with focus-visible,
+    // re-apply the class when the user switches back to the tab.
+    if (hadFocusVisibleRecently) {
+      hadKeyboardEvent = true;
     }
-  }, instance.focusVisibleCheckTime);
+  }
 }
 
-// The keys that might change document.activeElement.
-const FOCUS_KEYS = [
-  9, // 'Tab',
-  13, // 'Enter',
-  27, // 'Escape',
-  32, // ' ',
-  36, // 'Home',
-  35, // 'End',
-  37, // 'ArrowLeft',
-  38, // 'ArrowUp',
-  39, // 'ArrowRight',
-  40, // 'ArrowDown',
-];
+export function prepare() {
+  document.addEventListener('keydown', handleKeyDown, true);
+  document.addEventListener('mousedown', handlePointerDown, true);
+  document.addEventListener('pointerdown', handlePointerDown, true);
+  document.addEventListener('touchstart', handlePointerDown, true);
+  document.addEventListener('visibilitychange', handleVisibilityChange, true);
+}
 
-function isFocusKey(event) {
-  // Use event.keyCode to support IE 11
-  return FOCUS_KEYS.indexOf(event.keyCode) > -1;
+export function teardown() {
+  document.removeEventListener('keydown', handleKeyDown, true);
+  document.removeEventListener('mousedown', handlePointerDown, true);
+  document.removeEventListener('pointerdown', handlePointerDown, true);
```

Comment:
Is the pointerdown event redundant with mousedown and touchstart?

Reply:
I haven't really looked into those. Not sure what other pointer types e.g. a pen would trigger in browsers that implement PointerEvents.

Reply:
Keeping it for now until we can verify that a pointerdown from a pen triggers either keydown or mousedown.